scenes/enemies/guard/guard.gd

```diff
@@ -0,0 +1,395 @@
+## means the player is in sight.
+@onready var detection_area: Area2D = %DetectionArea
+## Progress bar that indicates how aware the guard is of the player, if it
+## is completely filled, [signal player_detected] is triggered.
+@onready var player_awareness: TextureProgressBar = %PlayerAwareness
+## Smaller detection area that only detectes when the player is really
+## close to the guard. If it detects a player, [signal player_detected] is
+## triggered.
+@onready var instant_detection_area: Area2D = %InstantDetectionArea
+## RayCast used to detect if the sight to a position is blocked.
+@onready var sight_ray_cast: RayCast2D = %SightRayCast
+## Control to hold debug info that can be toggled on or off.
+@onready var debug_info: Label = %DebugInfo
+## Handles the velocity and movement of the guard.
+@onready var guard_movement: GuardMovement = %GuardMovement
+
+
+func _ready() -> void:
+	if not Engine.is_editor_hint():
+		# Player awareness is configured and started empty.
+		if player_awareness:
+			player_awareness.max_value = time_to_detect_player
+			player_awareness.value = 0.0
+
+	if detection_area:
+		detection_area.scale = Vector2.ONE * detection_area_scale
+
+	# When the level starts, the guard is placed at the beginning of the
+	# patrol path.
+	if patrol_path:
+		global_position = _patrol_point_position(0)
+
+	guard_movement.destination_reached.connect(self._on_destination_reached)
+	guard_movement.still_time_finished.connect(self._on_still_time_finished)
+	guard_movement.path_blocked.connect(self._on_path_blocked)
+
+
+func _process(delta: float) -> void:
+	_update_debug_info()
+
+	if Engine.is_editor_hint() and not move_while_in_editor:
+		return
+
+	_process_state()
+	guard_movement.move()
+
+	_update_direction(delta)
+
+	var player_in_sight: Node2D = _player_in_sight()
+	_detect_player(player_in_sight)
+	_update_player_awareness(player_in_sight, delta)
+
+	_update_animation()
+
+
+## Updates the guard's movement behavior based on its current state.
+func _process_state() -> void:
+	match state:
+		State.PATROLING:
+			if patrol_path:
+				var target_position: Vector2 = _patrol_point_position(current_patrol_point_idx)
+				guard_movement.set_destination(target_position)
+			else:
+				guard_movement.stop_moving()
+		State.INVESTIGATING:
+			guard_movement.set_destination(last_seen_position)
+		State.DETECTING:
+			guard_movement.stop_moving()
+			if not _player_in_sight():
+				_change_state(State.INVESTIGATING)
+		State.RETURNING:
+			if not breadcrumbs.is_empty():
+				var target_position: Vector2 = breadcrumbs.back()
+				guard_movement.set_destination(target_position)
+			else:
+				_change_state(State.PATROLING)
+		State.ALERTED:
+			guard_movement.stop_moving()
+
+
+## Updates where the Guard is looking at.
+func _update_direction(delta: float) -> void:
+	if velocity.is_zero_approx():
+		return
+
+	var target_angle: float = velocity.angle()
+	var turn_speed: float = 10.0
```

Comment:
This `turn_speed` doesn't change, so it could be a `const` at the top of the script.